File: src/TypeFormatter/UnionTypeFormatter.ts
import type { JSONSchema7 } from "json-schema";
import type { Definition } from "../Schema/Definition.js";
import type { SubTypeFormatter } from "../SubTypeFormatter.js";
import type { BaseType } from "../Type/BaseType.js";
import { LiteralType } from "../Type/LiteralType.js";
import { NeverType } from "../Type/NeverType.js";
import { UnionType } from "../Type/UnionType.js";
import type { TypeFormatter } from "../TypeFormatter.js";
import { derefType } from "../Utils/derefType.js";
import { getTypeByKey } from "../Utils/typeKeys.js";
import { uniqueArray } from "../Utils/uniqueArray.js";
import { JsonTypeError } from "../Error/Errors.js";

type DiscriminatorType = "json-schema" | "open-api";

export class UnionTypeFormatter implements SubTypeFormatter {
    public constructor(
        protected childTypeFormatter: TypeFormatter,
        private discriminatorType?: DiscriminatorType,
    ) {}

    public supportsType(type: BaseType): boolean {
        return type instanceof UnionType;
    }
    private getTypeDefinitions(type: UnionType) {
        return type
            .getTypes()
            .filter((item) => !(derefType(item) instanceof NeverType))
            .map((item) => this.childTypeFormatter.getDefinition(item));
    }

    private getJsonSchemaDiscriminatorDefinition(type: UnionType): Definition {
        const definitions = this.getTypeDefinitions(type);
        const discriminator = type.getDiscriminator();

        if (!discriminator) {
            throw new JsonTypeError("discriminator is undefined", type);
        }

        const unionTypes = type.getTypes().filter((item) => !(derefType(item) instanceof NeverType));
        const kindTypes = unionTypes.map((item) => getTypeByKey(item, new LiteralType(discriminator)));

        // Separate types with and without discriminator field (non-congruent support)
        const typesWithDiscriminator: { kindType: BaseType; definition: Definition }[] = [];
        const typesWithoutDiscriminator: { definition: Definition }[] = [];

        for (let i = 0; i < kindTypes.length; i++) {
            if (kindTypes[i] === undefined) {
                // Type doesn't have discriminator field
                typesWithoutDiscriminator.push({ definition: definitions[i] });
            } else {
                typesWithDiscriminator.push({
                    kindType: kindTypes[i] as BaseType,
                    definition: definitions[i],
                });
            }
        }

        const kindDefinitions = typesWithDiscriminator.map((item) =>
            this.childTypeFormatter.getDefinition(item.kindType),
        );

        const allOf = [];

        // Add conditional schemas for types WITH discriminator field
        for (let i = 0; i < typesWithDiscriminator.length; i++) {
            allOf.push({
                if: {
                    properties: { [discriminator]: kindDefinitions[i] },
                    required: [discriminator],
                },
                then: typesWithDiscriminator[i].definition,
            });
        }

        // Add conditional schemas for types WITHOUT discriminator field (non-congruent)
        if (typesWithoutDiscriminator.length > 0) {
            allOf.push({
                if: {
                    not: {
                        properties: { [discriminator]: {} },
                        required: [discriminator],
                    },
                },
                then:
                    typesWithoutDiscriminator.length === 1
                        ? typesWithoutDiscriminator[0].definition
                        : {
                              anyOf: typesWithoutDiscriminator.map((item) => item.definition),
                          },
            });
        }

        const kindValues = kindDefinitions
            .flatMap((item) => item.const ?? item.enum)
            .filter((item): item is string | number | boolean | null => item !== undefined);

        const duplicates = kindValues.filter((item, index) => kindValues.indexOf(item) !== index);
        if (duplicates.length > 0) {
            throw new JsonTypeError(
                `Duplicate discriminator values: ${duplicates.join(", ")} in type ${JSON.stringify(type.getName())}.`,
                type,
            );
        }

        // Build properties only if we have types with discriminator
        const properties =
            kindValues.length > 0
                ? {
                      [discriminator]: {
                          enum: kindValues,
                      },
                  }
                : {};

        // Only require discriminator if all types have it
        const required = typesWithoutDiscriminator.length === 0 ? [discriminator] : [];

        return { type: "object", properties, required, allOf };
    }
    private getOpenApiDiscriminatorDefinition(type: UnionType): Definition {
        const oneOf = this.getTypeDefinitions(type);
        const discriminator = type.getDiscriminator();

        if (!discriminator) {
            throw new JsonTypeError("discriminator is undefined", type);
        }

        return {
            type: "object",
            discriminator: { propertyName: discriminator },
            required: [discriminator],
            oneOf,
        } as JSONSchema7;
    }
    public getDefinition(type: UnionType): Definition {
        const discriminator = type.getDiscriminator();
        if (discriminator !== undefined) {
            if (this.discriminatorType === "open-api") return this.getOpenApiDiscriminatorDefinition(type);
            return this.getJsonSchemaDiscriminatorDefinition(type);
        }

        const definitions = this.getTypeDefinitions(type);

        const flattenedDefinitions: JSONSchema7[] = [];

        // Flatten anyOf inside anyOf unless the anyOf has an annotation
        for (const def of definitions) {
            const keys = Object.keys(def);

            if (keys.length === 1 && keys[0] === "anyOf") {
                flattenedDefinitions.push(...(def.anyOf as any));
            } else {
                flattenedDefinitions.push(def);
            }
        }

        return flattenedDefinitions.length > 1
            ? {
                  anyOf: flattenedDefinitions,
              }
            : flattenedDefinitions[0];
    }
    public getChildren(type: UnionType): BaseType[] {
        return uniqueArray(
            type
                .getTypes()
                .reduce((result: BaseType[], item) => [...result, ...this.childTypeFormatter.getChildren(item)], []),
        );
    }
}

File: internal/formatter/union_type_formatter.go
package formatter

import (
	"fmt"
	"strings"

	"github.com/vega/ts-json-schema-generator/internal/config"
	"github.com/vega/ts-json-schema-generator/internal/schema"
	"github.com/vega/ts-json-schema-generator/internal/types"
)

// UnionTypeFormatter mirrors src/TypeFormatter/UnionTypeFormatter.ts.
type UnionTypeFormatter struct {
	childTypeFormatter TypeFormatter
	discriminatorType  config.DiscriminatorType
}

func NewUnionTypeFormatter(childTypeFormatter TypeFormatter, discriminatorType config.DiscriminatorType) *UnionTypeFormatter {
	return &UnionTypeFormatter{childTypeFormatter: childTypeFormatter, discriminatorType: discriminatorType}
}

func (f *UnionTypeFormatter) SupportsType(t types.Type) bool {
	_, ok := t.(*types.UnionType)
	return ok
}

func (f *UnionTypeFormatter) getTypeDefinitions(unionType *types.UnionType) []*schema.Definition {
	definitions := []*schema.Definition{}
	for _, item := range unionType.Types() {
		if types.IsNeverLike(types.DerefType(item)) {
			continue
		}
		definitions = append(definitions, f.childTypeFormatter.GetDefinition(item))
	}
	return definitions
}

func (f *UnionTypeFormatter) getJSONSchemaDiscriminatorDefinition(unionType *types.UnionType) *schema.Definition {
	definitions := f.getTypeDefinitions(unionType)
	discriminator := unionType.Discriminator

	if discriminator == "" {
		panic(fmt.Errorf("discriminator is undefined in type %s", unionType.Name()))
	}

	var kindTypes []types.Type
	for _, item := range unionType.Types() {
		if types.IsNeverLike(types.DerefType(item)) {
			continue
		}
		kindTypes = append(kindTypes, types.GetTypeByKey(item, &types.LiteralType{Value: discriminator}))
	}

	for i, item := range kindTypes {
		if item == nil {
			panic(fmt.Errorf(
				"cannot find discriminator keyword %q in type %s",
				discriminator,
				unionType.Types()[i].Name(),
			))
		}
	}

	kindDefinitions := make([]*schema.Definition, len(kindTypes))
	for i, item := range kindTypes {
		kindDefinitions[i] = f.childTypeFormatter.GetDefinition(item)
	}

	allOf := []*schema.Definition{}
	for i := range definitions {
		ifProperties := schema.NewProperties()
		ifProperties.Set(discriminator, kindDefinitions[i])
		allOf = append(allOf, &schema.Definition{
			If:   &schema.Definition{Properties: ifProperties},
			Then: definitions[i],
		})
	}

	kindValues := []any{}
	for _, item := range kindDefinitions {
		if item.Const != nil {
			kindValues = append(kindValues, *item.Const)
		} else if item.Enum != nil {
			kindValues = append(kindValues, item.Enum...)
		}
	}

	var duplicates []any
	for i, item := range kindValues {
		for j := 0; j < i; j++ {
			if kindValues[j] == item {
				duplicates = append(duplicates, item)
				break
			}
		}
	}
	if len(duplicates) > 0 {
		strs := make([]string, len(duplicates))
		for i, d := range duplicates {
			strs[i] = rawValueToString(d)
		}
		panic(fmt.Errorf(
			"duplicate discriminator values: %s in type %q",
			strings.Join(strs, ", "),
			unionType.Name(),
		))
	}

	properties := schema.NewProperties()
	properties.Set(discriminator, &schema.Definition{Enum: kindValues})

	return &schema.Definition{
		Type:       "object",
		Properties: properties,
		Required:   []string{discriminator},
		AllOf:      allOf,
	}
}

func (f *UnionTypeFormatter) getOpenAPIDiscriminatorDefinition(unionType *types.UnionType) *schema.Definition {
	oneOf := f.getTypeDefinitions(unionType)
	discriminator := unionType.Discriminator

	if discriminator == "" {
		panic(fmt.Errorf("discriminator is undefined in type %s", unionType.Name()))
	}

	return &schema.Definition{
		Type:          "object",
		Discriminator: map[string]any{"propertyName": discriminator},
		Required:      []string{discriminator},
		OneOf:         oneOf,
	}
}

func (f *UnionTypeFormatter) GetDefinition(t types.Type) *schema.Definition {
	unionType := t.(*types.UnionType)

	if unionType.Discriminator != "" {
		if f.discriminatorType == config.DiscriminatorOpenAPI {
			return f.getOpenAPIDiscriminatorDefinition(unionType)
		}
		return f.getJSONSchemaDiscriminatorDefinition(unionType)
	}

	definitions := f.getTypeDefinitions(unionType)

	// Flatten anyOf inside anyOf unless the anyOf has an annotation.
	flattenedDefinitions := []*schema.Definition{}
	for _, def := range definitions {
		if def.AnyOf != nil && isOnlyAnyOf(def) {
			flattenedDefinitions = append(flattenedDefinitions, def.AnyOf...)
		} else {
			flattenedDefinitions = append(flattenedDefinitions, def)
		}
	}

	if len(flattenedDefinitions) > 1 {
		return &schema.Definition{AnyOf: flattenedDefinitions}
	}
	if len(flattenedDefinitions) == 1 {
		return flattenedDefinitions[0]
	}
	// The TypeScript implementation returns undefined here; an empty
	// definition is the closest equivalent.
	return &schema.Definition{}
}

func (f *UnionTypeFormatter) GetChildren(t types.Type) []types.Type {
	var children []types.Type
	for _, item := range t.(*types.UnionType).Types() {
		children = append(children, f.childTypeFormatter.GetChildren(item)...)
	}
	return uniqueTypes(children)
}

// isOnlyAnyOf reports whether anyOf is the definition's only key
// (`Object.keys(def).length === 1 && keys[0] === "anyOf"`).
func isOnlyAnyOf(def *schema.Definition) bool {
	check := *def
	check.AnyOf = nil
	return isEmptyDefinition(&check)
}

// rawValueToString renders a raw JSON value the way JavaScript's implicit
// string conversion does in Array#join.
func rawValueToString(v any) string {
	switch x := v.(type) {
	case nil:
		return "null"
	case string:
		return x
	case bool:
		if x {
			return "true"
		}
		return "false"
	case float64:
		return types.NumberToString(x)
	case int:
		return types.NumberToString(float64(x))
	default:
		return fmt.Sprintf("%v", x)
	}
}
